applications/ConvectionDiffusionApplication/custom_response_functions/local_temperature_average_response_function.h

```diff
@@ -160,8 +160,26 @@ class LocalTemperatureAverageResponseFunction: public AdjointResponseFunction
     double CalculateValue(ModelPart& rModelPart) override
     {
         KRATOS_TRY;
-        KRATOS_ERROR
-            << "PointTemperature::CalculateValue(ModelPart& rModelPart) is not implemented!!!\n";
+
+        const Communicator& r_communicator = rModelPart.GetCommunicator();
+        const int number_of_nodes = r_communicator.LocalMesh().NumberOfNodes();
+        
+	double domain_aggregated_temperature = 0.0;
+
+#pragma omp parallel for reduction(+ : domain_aggregated_temperature)
```

Comment:
is it necessary to consider the target model part?

```cpp
        std::string target_model_part = Settings["model_part_name"].GetString();

        auto& r_target_model_part = GetTargetModelPart(rModelPart, target_model_part);
        auto& r_nodes = r_target_model_part.Nodes();
        mNumNodes = r_nodes.size();

        VariableUtils variable_utils;
        variable_utils.SetFlag(STRUCTURE,true,r_nodes);
```

Reply:
See constructor of the response. I think these are a subset of nodes where the actual temperature is considered

Reply:
Thats what I am not sure about, would the input arg ModelPart will have target model part as a submodel part or is it the target model part? What is being passed by the shape optimization? Does it pass the whole model part, or the model part, which is being optimized?

If the latter case, then we need to store target model part in the response function as I think.

Reply:
Unfortunately it is even more complicated:

The adjoint response function is created with the adjoint model part.
However the primal model part is passed to the CalculateValue function. 

For this response as simply nodal values are summed up, we could use both to determine the response value (because the variables are transferred beforehands), but in many cases the primal model part has to be used.

To be correct, we would need to store the name of the target model part and then get it from the model_part that is passed to the response. The flags set for the nodes in the constructor will not be there as thei are on the adjoint model part...